src/rsz/src/SpefWriter.cc

```diff
@@ -0,0 +1,246 @@
+// POSSIBILITY OF SUCH DAMAGE.
+//
+///////////////////////////////////////////////////////////////////////////////
+
+#include "rsz/SpefWriter.hh"
+
+#include <fstream>
+#include <iostream>
+
+#include "db_sta/dbNetwork.hh"
+#include "sta/Corner.hh"
+#include "sta/Parasitics.hh"
+#include "sta/Units.hh"
+#include "utl/Logger.h"
+
+namespace rsz {
+
+using std::map;
+using utl::RSZ;
+
+SpefWriter::SpefWriter(Logger* logger,
+                       dbSta* sta,
+                       std::map<Corner*, std::ostream*>& spef_streams)
+    : logger_(logger),
+      sta_(sta),
+      network_(sta_->getDbNetwork()),
+      parasitics_(sta_->parasitics()),
+      spef_streams_(spef_streams)
+{
+}
+
+void SpefWriter::writeSpefHeader(Corner* corner)
+{
+  auto it = spef_streams_.find(corner);
+  if (it == spef_streams_.end()) {
+    return;
+  }
+  std::ostream& stream = *it->second;
+  stream << "*SPEF \"ieee 1481-1999\"" << '\n';
+  stream << "*DESIGN \"" << network_->block()->getName() << "\"" << '\n';
+  stream << "*DATE \"11:11:11 Fri 11 11, 1111\"" << '\n';
+  stream << "*VENDOR \"The OpenROAD Project\"" << '\n';
+  stream << "*PROGRAM \"OpenROAD\"" << '\n';
+  stream << "*VERSION \"1.0\"" << '\n';
+  stream << "*DESIGN_FLOW \"NAME_SCOPE LOCAL\" \"PIN_CAP NONE\"" << '\n';
+  stream << "*DIVIDER /" << '\n';
+  stream << "*DELIMITER :" << '\n';
+  stream << "*BUS_DELIMITER []" << '\n';
+
+  auto units = network_->units();
+  std::string time_unit = std::string(units->timeUnit()->suffix());
+  std::string cap_unit = std::string(units->capacitanceUnit()->suffix());
+  std::string res_unit = std::string(units->resistanceUnit()->suffix());
+  std::transform(
+      time_unit.begin(), time_unit.end(), time_unit.begin(), ::toupper);
+  std::transform(cap_unit.begin(), cap_unit.end(), cap_unit.begin(), ::toupper);
+  std::transform(res_unit.begin(), res_unit.end(), res_unit.begin(), ::toupper);
+
+  stream << "*T_UNIT 1 " << time_unit << '\n';
+  stream << "*C_UNIT 1 " << cap_unit << '\n';
+  stream << "*R_UNIT 1 " << res_unit << '\n';
+  stream << '\n';
+}
+
+void SpefWriter::writeSpefPorts(Corner* corner)
+{
+  auto it = spef_streams_.find(corner);
+  if (it == spef_streams_.end()) {
+    return;
+  }
+  std::ostream& stream = *it->second;
+
+  std::unique_ptr<sta::InstancePinIterator> pin_iter(
+      network_->pinIterator(network_->topInstance()));
+
+  stream << "*PORTS" << '\n';
+  while (pin_iter->hasNext()) {
+    sta::Pin* pin = pin_iter->next();
+    odb::dbITerm* iterm = nullptr;
+    odb::dbBTerm* bterm = nullptr;
+    odb::dbModITerm* moditerm = nullptr;
+    odb::dbModBTerm* modbterm = nullptr;
+    network_->staToDb(pin, iterm, bterm, moditerm, modbterm);
+
+    if (iterm != nullptr) {
+      stream << iterm->getName() << " ";
+      if (iterm->getIoType() == odb::dbIoType::INPUT) {
+        stream << "I";
+      } else if (iterm->getIoType() == odb::dbIoType::OUTPUT) {
+        stream << "O";
+      } else {
+        stream << "B";
+      }
```

Comment:
Stylistically I would avoid repeating this block and just get the io type in the iterm/bterm blocks and then convert it to text once afterwards.